Specialize Iterator for &mut I where I: Sized

SkiFire13 opened the pull request:
This PR aims to improve the implementation of `Iterator` and `DoubleEndedIterator` for `&mut I where I: Iterator`. This is accomplished by forwarding `try_fold` to the implementation for `I` (since it takes `&mut self`) and implementing `fold` in terms of `try_fold` (since `fold` takes `self`).

Since `try_fold` is not available if `I: !Sized` I needed to specialize the case `I: Sized` (that's also why I choose to keep the `+ Sized` bound, to be explicit about what I'm specializing). This means this optimization won't apply to `&mut dyn Iterator`.

While coding this pr I realized that it adds a considerable chunk of code for `&mut I where I: Iterator` and it may make sense to move it to a separate file just like other adapters. Tell me what you think about it.

Edit: Note that the following benchmark results are relatively old (they were done before the LLVM 12 update for instance) and should probably be updated. 

Here's a comparison (run with `cargo benchcmp master pr --threshold 2`) of the benchmark (run with `py x.py bench --stage 0 library/core --test-args iter::`):

```
 name                                         master.txt ns/iter  pr.txt ns/iter  diff ns/iter   diff %  speedup
 iter::bench_cycle_skip_take_ref_sum          2,702,500           1,037,230         -1,665,270  -61.62%   x 2.61
 iter::bench_cycle_take_ref_sum               950,040             738,915             -211,125  -22.22%   x 1.29
 iter::bench_cycle_take_skip_ref_sum          2,117,830           666,940           -1,450,890  -68.51%   x 3.18
 iter::bench_enumerate_chain_ref_sum          1,381,700           1,476,370             94,670    6.85%   x 0.94
 iter::bench_enumerate_ref_sum                459,605             436,760              -22,845   -4.97%   x 1.05
 iter::bench_filter_chain_ref_count           1,389,580           1,095,382           -294,198  -21.17%   x 1.27
 iter::bench_filter_chain_ref_sum             2,007,972           893,280           -1,114,692  -55.51%   x 2.25
 iter::bench_filter_map_chain_ref_sum         3,908,650           1,040,480         -2,868,170  -73.38%   x 3.76
 iter::bench_filter_map_ref_sum               806,280             469,225             -337,055  -41.80%   x 1.72
 iter::bench_filter_ref_sum                   604,827             566,450              -38,377   -6.35%   x 1.07
 iter::bench_flat_map_chain_ref_sum           4,467,065           1,641,630         -2,825,435  -63.25%   x 2.72
 iter::bench_flat_map_ref_sum                 1,849,160           474,605           -1,374,555  -74.33%   x 3.90
 iter::bench_flat_map_sum                     446,975             483,936               36,961    8.27%   x 0.92
 iter::bench_for_each_chain_ref_fold          1,608,805           1,378,055           -230,750  -14.34%   x 1.17
 iter::bench_fuse_chain_ref_sum               1,612,977           942,280             -670,697  -41.58%   x 1.71
 iter::bench_fuse_chain_sum                   461,980             707,125              245,145   53.06%   x 0.65
 iter::bench_fuse_ref_sum                     460,043             352,918             -107,125  -23.29%   x 1.30
 iter::bench_inspect_chain_ref_sum            1,612,792           941,940             -670,852  -41.60%   x 1.71
 iter::bench_inspect_chain_sum                460,000             706,720              246,720   53.63%   x 0.65
 iter::bench_inspect_ref_sum                  459,523             358,241             -101,282  -22.04%   x 1.28
 iter::bench_multiple_take                    63                  58                        -5   -7.94%   x 1.09
 iter::bench_peekable_chain_ref_sum           1,378,620           924,195             -454,425  -32.96%   x 1.49
 iter::bench_peekable_chain_sum               463,480             692,655              229,175   49.45%   x 0.67
 iter::bench_peekable_ref_sum                 459,677             346,225             -113,452  -24.68%   x 1.33
 iter::bench_rposition                        66                  74                         8   12.12%   x 0.89
 iter::bench_skip_chain_ref_sum               3,911,640           921,185           -2,990,455  -76.45%   x 4.25
 iter::bench_skip_cycle_skip_zip_add_ref_sum  4,439,160           3,252,870         -1,186,290  -26.72%   x 1.36
 iter::bench_skip_sum                         344,301             460,145              115,844   33.65%   x 0.75
 iter::bench_skip_while_chain_ref_sum         3,947,070           921,160           -3,025,910  -76.66%   x 4.28
 iter::bench_skip_while_ref_sum               344,387             460,540              116,153   33.73%   x 0.75
 iter::bench_take_while_chain_ref_sum         760,888             511,980             -248,908  -32.71%   x 1.49
```

There are a lot of promising improvements, but surprisingly there are also a couple of strange regressions:

```
 name                                         master.txt ns/iter  pr.txt ns/iter  diff ns/iter   diff %  speedup
 iter::bench_flat_map_sum                     446,975             483,936               36,961    8.27%   x 0.92
 iter::bench_fuse_chain_sum                   461,980             707,125              245,145   53.06%   x 0.65
 iter::bench_inspect_chain_sum                460,000             706,720              246,720   53.63%   x 0.65
 iter::bench_peekable_chain_sum               463,480             692,655              229,175   49.45%   x 0.67
 iter::bench_rposition                        66                  74                         8   12.12%   x 0.89
 iter::bench_skip_sum                         344,301             460,145              115,844   33.65%   x 0.75
```

These benchmarks shouldn't have be impacted by my changes because they never use `&mut impl Iterator` but somehow they were. My guess is that this is the result of some LLVM shenanigans because if I remove any benchmark containing `.by_ref()` then most of them don't regress anymore. Since these look like codegen problems and specific to the current benchmark I'm not too worried about them.

```
 name                                         master.txt ns/iter  pr.txt ns/iter  diff ns/iter   diff %  speedup
 iter::bench_enumerate_chain_ref_sum          1,381,700           1,476,370             94,670    6.85%   x 0.94
 iter::bench_skip_while_ref_sum               344,387             460,540              116,153   33.73%   x 0.75
```

These benchmarks however were directly impacted by my changes and the give the same results even if I remove all the other benchmarks, so I guess they're genuine regressions. 

All these regressions are present even if I copy the default `fold` implementation but directly call `I::next` instead of passing through `<&mut I>::next` (which should just call `I::next` anyway). This makes me think these are also codegen problem.

I would like to not have these regressions but the other improvements probably outweights them.

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

pickfire commented on library/core/src/iter/traits/double_ended.rs:
```diff
@@ -360,4 +360,77 @@ impl<'a, I: DoubleEndedIterator + ?Sized> DoubleEndedIterator for &'a mut I {
+    where
+        F: FnMut(B, Self::Item) -> B,
+    {
+        #[inline]
+        fn ok<T, U>(mut f: impl FnMut(T, U) -> T) -> impl FnMut(T, U) -> Result<T, !> {
+            move |acc, x| Ok(f(acc, x))
+        }
```
Why can't we inline this?

SkiFire13 replied:
I just tried to be consistent with the rest of the stdlib.

bluss replied:
It's for avoiding generics bloat

pickfire commented on library/core/src/iter/traits/iterator.rs:
```diff
@@ -3394,4 +3394,77 @@ impl<I: Iterator + ?Sized> Iterator for &mut I {
+}
+
+impl<I: Iterator + Sized> SpecSizedIterator for &mut I {
+    #[inline]
+    fn try_fold<B, F, R>(&mut self, init: B, f: F) -> R
+    where
+        F: FnMut(B, Self::Item) -> R,
+        R: Try<Ok = B>,
+    {
+        (**self).try_fold(init, f)
+    }
```
Why do we need this part? It seemed similar to `default fn try_fold`, I don't understand this part.

SkiFire13 replied:
This calls the `try_fold` implementation for `I` instead of the default implementation provided to `&mut I`

pickfire replied:
Thanks for explaining, I didn't know about this. On how `**self` gets from `&mut I` to `I`.

SkiFire13 replied:
I guess it would be more clear if it was called like `<I as Iterator>::try_fold(self, init, f)`, but that's for another PR

the8472 commented on library/core/src/iter/traits/iterator.rs:
```diff
@@ -3394,4 +3394,77 @@ impl<I: Iterator + ?Sized> Iterator for &mut I {
+        while let Some(x) = self.next() {
+            accum = f(accum, x)?;
+        }
+        try { accum }
```
Could the copy-pasted functions be extracted into a free function which is then called by the `Iterator` and `SpecSizedIterator` implementations? That way the actual implementation only has to be maintained in one place.

pickfire replied:
Without the generics, nice. I didn't thought about this.

SkiFire13 replied:
Nice idea! I think it still has to be generic though

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1912815450) [(plain)](https://github.com/rust-lang/rust/commit/665e27934252ba8b5390d82e906ff4cf1dff98c3/checks/1912815450/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9200/11455
.................................................................................................... 9300/11455
.................................................................................................... 9400/11455
..............i......i.............................................................................. 9500/11455
....................................................iiiiiii..iiiiii.i............................... 9600/11455
.................................................................................................... 9800/11455
.................................................................................................... 9900/11455
.................................................................................................... 10000/11455
.................................................................................................... 10100/11455
---

---- [ui] ui/issues/issue-3044.rs stdout ----
diff of stderr:

11 note: associated function defined here
12   --> $SRC_DIR/core/src/iter/traits/iterator.rs:LL:COL
13    |
- LL |     fn fold<B, F>(mut self, init: B, mut f: F) -> B
+ LL |     fn fold<B, F>(self, init: B, f: F) -> B
16 
17 error: aborting due to previous error



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-3044/issue-3044.stderr
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args issues/issue-3044.rs`
error: 1 errors occurred comparing output.
status: exit code: 1
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/issues/issue-3044.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-3044" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-3044/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error[E0061]: this function takes 2 arguments but 1 argument was supplied
  --> /checkout/src/test/ui/issues/issue-3044.rs:3:23
   |
LL |       needlesArr.iter().fold(|x, y| {
   |  _______________________^^^^_-
   | |                       expected 2 arguments
LL | |     });
   | |_____- supplied 1 argument
   |
   |
note: associated function defined here
  --> /checkout/library/core/src/iter/traits/iterator.rs:2113:8
   |
LL |     fn fold<B, F>(self, init: B, f: F) -> B

error: aborting due to previous error

For more information about this error, try `rustc --explain E0061`.
---
test result: FAILED. 11361 passed; 1 failed; 93 ignored; 0 measured; 0 filtered out; finished in 137.81s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-9/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "9.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver engine executionengine fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:15:03
```

</details>

bluss commented:
Is the intermediate trait SpecSizedIterator a workaround? Could it not be specialized on the Sized bound for the plain `Iterator` trait impl for `&mut I`. I know I tried that years ago (and at that point it ICE'd). Just specializing the Iterator impl would avoid having to duplicate the default method implementations, but I admit I don't know the min specialization rules, if they allow specializing on `I: Sized`.

the8472 commented:
Current std policy says to use private helper traits. https://std-dev-guide.rust-lang.org/code-considerations/using-unstable-lang/specialization.html

Mark-Simulacrum commented:
r? @cuviper (or someone else more familiar with iterator code)

cuviper commented:
Let's check the general perf effect:

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit a09399b1f1ba90d72fdc4eadf65d0ab459a96287 with merge 4bf6a12cc7384e44f61e55b762c4efb6612c5aca...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a09399b1f1ba90d72fdc4eadf65d0ab459a96287","merge_sha":"4bf6a12cc7384e44f61e55b762c4efb6612c5aca"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1944324092)
Build commit: 4bf6a12cc7384e44f61e55b762c4efb6612c5aca (`4bf6a12cc7384e44f61e55b762c4efb6612c5aca`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1944324092"},"merge_sha":"4bf6a12cc7384e44f61e55b762c4efb6612c5aca"} -->

rust-timer commented:
Queued 4bf6a12cc7384e44f61e55b762c4efb6612c5aca with parent d2b38d6b3c9d1ee52a360c3ce61e54b7aa91d405, future [comparison URL](https://perf.rust-lang.org/compare.html?start=d2b38d6b3c9d1ee52a360c3ce61e54b7aa91d405&end=4bf6a12cc7384e44f61e55b762c4efb6612c5aca).

rust-timer commented:
Finished benchmarking try commit (4bf6a12cc7384e44f61e55b762c4efb6612c5aca): [comparison url](https://perf.rust-lang.org/compare.html?start=d2b38d6b3c9d1ee52a360c3ce61e54b7aa91d405&end=4bf6a12cc7384e44f61e55b762c4efb6612c5aca).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

cuviper commented:
@SkiFire13 those perf results don't look promising -- what's your take on it?

SkiFire13 commented:
I'm not quite sure but my guesses are:

* Calling `I::try_fold` instead of just calling `next` in a loop may result in a bigger function, which probably translates in bigger compile times. Since `println` somehow ends up calling `for_each` on a `&mut vec::Drain` (through `BufWriter::flush_buf` and then `vec::Drain::drop`) this might explain why the `incr-patched: println` builds were the most affected. Maybe we could change `vec::Drain::drop` to not call that `for_each`, or maybe discard the idea of specializing `&mut Iterator` and instead add another iterator adaptor that does the same job but is opt-in (this feels like duplicating `by_ref` but it's not like we can change its return type). 

* `default_fold` and friends being separate functions may result in more time spent deciding whether to inline them or not. I don't know exactly how this may affect compile times, but we may try manually inling (e.g. reverting the commit that introduced `default_fold`) or using `#[inline(always)]`. As a side note, a quick benchmark shows that after adding `default_fold` and friends some benchmarks didn't regress anymore, but others started regressing, which is pretty weird to me.

crlf0710 commented:
@SkiFire13 Ping from triage, what's next steps here?

SkiFire13 commented:
> @SkiFire13 Ping from triage, what's next steps here?

After the perf run I decided to focus on solving the regressing benchmarks, hoping that their cause is the same as the regressions in the perf run. I tried inling more methods and changing `vec::Drain::drop` to not use `.by_ref()` but unfortunately the result in both cases was that some benches regress and other improve. Now I'm a bit lost. I'm not too familiar with how codegen works and how to debug it and the fact that the regressions behaves pretty weirdly doesn't help at all.

the8472 commented:
If you click on the percentages you'll see a breakdown by what's consuming more time. It's mostly spending more time in LLVM, so one thing you could investigate is whether these changes caused more llvm IR to be emitted for those benchmarked crates. [cargo llvm-lines](https://github.com/dtolnay/cargo-llvm-lines) can be useful there.

In the past optimization attempts in `Vec` have shown that while internal iteration methods such as `fold()` can generate better code they involve a lot more duplicate IR due to all the helper functions in the iterator adapters which makes compilation more expensive than the speed gains in the compiler itself. E.g. a 3rd party crate might become faster when executed (-0.2%)  but the compiler spends more time in LLVM (+0.4%) and its own rust code only gets a little more efficient ( -0.1%) since it doesn't use that particular code path often which means it shows a +0.3% regression. Perf runs only capture the +0.3%, not the -0.2%.

> Since println somehow ends up calling for_each on a &mut vec::Drain (through BufWriter::flush_buf and then vec::Drain::drop) this might explain why the incr-patched: println builds were the most affected. 

Just run the experiment?

Another idea would be measuring the shuffling of the default implementations into separate traits (`Spec...`) and the additional `Iterator + Sized` specialization separately. Then we could see whether it's the outlining or the specialization of interest that causes the regression.

SkiFire13 commented:
Nothing really relevant. The only thing is that the benches in the first post are a bit outdated since they were made before the LLVM 12 update. Some of the benches that regress with my PR also regressed with the LLVM 12 update, so now technically there are less bench regressions I guess? I don't expect any change in a perf run though.

> Just run the experiment?

Just saw this, sorry if I'm a bit late. I already tried that locally and as I said in my previous comment some benches regressed, while others improved. Or do you mean to do a perf run with that change?

the8472 commented:
Yeah, meant a perf run with that change.

SkiFire13 commented:
I've commited that change, now it's ready for the perf run

cuviper commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 03cf7548e418587965e99ac2e45ed1ab0c51ba90 with merge d1fceadf2e329af21b508486e005586fb36e8466...
<!-- homu: {"type":"TryBuildStarted","head_sha":"03cf7548e418587965e99ac2e45ed1ab0c51ba90","merge_sha":"d1fceadf2e329af21b508486e005586fb36e8466"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2309652568)
Build commit: d1fceadf2e329af21b508486e005586fb36e8466 (`d1fceadf2e329af21b508486e005586fb36e8466`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2309652568"},"merge_sha":"d1fceadf2e329af21b508486e005586fb36e8466"} -->

rust-timer commented:
Queued d1fceadf2e329af21b508486e005586fb36e8466 with parent da0b9b6485585a3557e882acbf5054e826625031, future [comparison URL](https://perf.rust-lang.org/compare.html?start=da0b9b6485585a3557e882acbf5054e826625031&end=d1fceadf2e329af21b508486e005586fb36e8466).

rust-timer commented:
Finished benchmarking try commit (d1fceadf2e329af21b508486e005586fb36e8466): [comparison url](https://perf.rust-lang.org/compare.html?start=da0b9b6485585a3557e882acbf5054e826625031&end=d1fceadf2e329af21b508486e005586fb36e8466).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

SkiFire13 commented:
Ouch, that's worse than before. Now I'm really out of ideas

the8472 commented:
I don't quite understand what the drop implementation of Drain is doing anyway. It loops, creates a drop guard and if that panics it tries to loop again inside drop guard, which could very well lead to a double panic? I haven't seen that being done anywhere else.

SkiFire13 commented:
It tries to not leak the other elements if dropping one of them panics. It should have been added in #67290 along with the same logic for other iterators

the8472 commented:
Interesting. I think the resulting Drain implementation is far from optimal. It could probably calculate slices and do some `ptr::drop_in_place(slice)` instead, which already does the continue-on-panic thing if I understand https://github.com/rust-lang/rust/pull/67290#pullrequestreview-332322049 correctly

JohnCSimon commented:
ping from triage: 
@SkiFire13 can you please post your status on this PR? 
thank you

SkiFire13 commented:
> 
> 
> Ouch, that's worse than before. Now I'm really out of ideas

@JognCSimon there unfortunately there has been no progress from this comment. I'm not sure how to proceed from here. If anyone has any idea, they're welcome, otherwise I might as well close this PR.

the8472 commented:
Well, the drain implementation still looks pretty bad to me. If that really causes a large chunk of the regressions then maybe it would make sense to optimize it in a separate PR and then rebase and retest this one once that goal is achieved.

SkiFire13 commented:
I've rebased onto #85157, let's see how it compares. It might be nice to have a comparison with both master and #85157. I've also converted this PR into a draft since it's clearly not ready for merging.

Mark-Simulacrum commented:
@bors try @rust-timer queue

bors commented:
:hourglass: Trying commit 17350deaf6c511ed778d15f592e0ff1917fd6c73 with merge 5d8726c5e6e2c8a635a0178c21932f6634271e89...
<!-- homu: {"type":"TryBuildStarted","head_sha":"17350deaf6c511ed778d15f592e0ff1917fd6c73","merge_sha":"5d8726c5e6e2c8a635a0178c21932f6634271e89"} -->

rust-timer commented:
Queued 5d8726c5e6e2c8a635a0178c21932f6634271e89 with parent 2bafe96272bf22cd191c6a01741184b60063292d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=2bafe96272bf22cd191c6a01741184b60063292d&end=5d8726c5e6e2c8a635a0178c21932f6634271e89).

Dylan-DPC commented:
@SkiFire13 any updates on this?

SkiFire13 commented:
I left this PR in draft mode in case I got new ideas but nothing came up and eventually I forgot about it.
